Create version of Sankey plot with an additional node and correct toy data set

lindadelacombaz opened the pull request:
Closes #2 
Dear @AnneSchoenauer,

I have addressed the [review you made](https://github.com/2DegreesInvesting/tiltPlot/issues/2#issuecomment-1603760464). Here are the points I have changed and addressed : 

> It should be in general possible to have the company nodes included. Maybe you could already implement this so that we have two versions of the sankey_plot?

1. I have added a new parameter called _with_company_ in the plot_sankey() function where now you have two options:
- plot_sankey(toy_data, with_company = TRUE) plots an additional node with the companies name, 
- plot_sankey(toy_data, with_company = FALSE) plots the sankey graph without the companies additional node like before.

>  If the company only has 1 distinct product, how can it be that it is allocated to two different sectors?

3. I have changed the toy data set, I did not understand correctly the toy data set and know it should make sense with the number of distinct products and sectors :).

> I wonder why the flows are smaller from sector to risk category as they are going from A/B to sectors. Shouldn't they be exactly the same as the bars should represent exactly the same amount of money before and after the sector split?

4. I also misunderstood the final Sankey plot so now the size of the flows are all the same and represents the amount of the money that the bank borrowed to the company. 

It would be great if you tested the new output by running in your **console** : 

devtools::install_github("2DegreesInvesting/tiltPlot@2_create_sankey_plot")

And run the three lines of code in the README file to check the new output of the Sankey plot :) 

Let me know if you need any help to do that.

Once that is done, I will change the PR as ready for review and ask @maurolepore for a review.

AnneSchoenauer commented:
Dear @lindadelacombaz,
From your written explanations, it sounds like that you incorporated all my concerns that I had with the plot. And how nice is the option with company=TRUE/FALSE! Well done! :) 

I installed the package again and then applied the code above. At the first time the following error message occured after opening the library(tilt_plot): 
```
Error: lazy-load database 'C:/Users/anne-/AppData/Local/R/win-library/4.2/tiltPlot/R/tiltPlot.rdb' is corrupt
In addition: Warning messages:
1: restarting interrupted promise evaluation 
2: internal error -3 in R_decompress1 
```
However, when I closed the library and re-opened it the code run smoothly. The only warning messages occured were: 
`> plot_sankey(toy_data, with_company = TRUE) 
Links is a tbl_df. Converting to a plain data frame.
Nodes is a tbl_df. Converting to a plain data frame.`

Content-wise the graphs looks really really great! One last "beauty" request.  If you see now the graph the orange flow is interrupted with the node. Would it be possible that the flows stop and start at the middle node at the same location? So that it is really one flow at the end? If this is not possible and is too much work don't wory about this!

<img width="293" alt="image" src="https://github.com/2DegreesInvesting/tiltPlot/assets/115623567/e62f04bb-2f83-4040-a411-a691f615654b">

Also, can we change colour-codes etc. at a later point? Also there might be also different nodes coming up in the future. Meaning we might add another node showing the probability of default or the loan amount. I assume that this is easily done or shall I already update the toy data set accordingly?

lindadelacombaz commented:
Dear @maurolepore, 
I have addressed some comments that Anne made. (c.f. [comment above](https://github.com/2DegreesInvesting/tiltPlot/pull/10#issue-1774790254)) and changed my code to address her review.

I made her give me feedback before merging using your great tip and she installed and tested my branch by running devtools::install_github("2DegreesInvesting/tiltPlot@2_create_sankey_plot").

The comment she made will be addressed in separates PR and tickets.

My code is ready to be merged - but wanted you to take a look at it before :).

Thank you !

maurolepore commented on R/plot_sankey.R:
```diff
@@ -1,37 +1,63 @@
+  }
 
   links <- data_links |>
     select(
       "bank",
       source = "middle_node2",
       target = "pctr_risk_category",
-      value = "amount_of_disctinct_products",
+      value = "amount",
       group = "pctr_risk_category"
     ) |>
     bind_rows(links)
```
Here again. This pattern appears 4 times in this function.

Note that most of this function deals with data preparation. But this is a plotting function so most of it should be about plotting. This suggests it'll be best to extract data preparation elsewhere. Not sure if you would export that so that the user does the preparation, or if you would call the helper from the plotting function. 

```r
prepare_data(data) |> plot_sankey()

# Or
plot_sankey <- function(...) {
  prepare_data(data) |> plot_sankey_impl()
}
plot_sankey(data)
```

You may see 2dii.plot for inspiration about design.

maurolepore commented on data-raw/toy_data.R:
```diff
@@ -1,10 +1,10 @@
-    "A",       5,                             1,               "low",    "sector_b",
-    "B",       2,                             2,               "medium", "sector_a",
-    "B",       10,                            2,               "high",   "sector_b",
+  ~bank, ~amount, ~company_name, ~amount_of_distinct_products, ~pctr_risk_category,  ~tilt_sec,
+  "A",       5,       "Peter",                            2,               "low", "sector_a",
+  "A",       5,       "Peter",                            2,              "high", "sector_b",
+  "B",      10,       "Mario",                            1,            "medium", "sector_c",
+  "C",       8,       "Mario",                            1,               "low", "sector_c"
```
This looks miss-aligned by 1-2 characters?

lindadelacombaz replied:
Oops, you're right. I'll change that in a separate PR so that Anne can test the code with the Bundesbank :) 
Thank you for your comments @maurolepore. I think the next step would be to transfer into a ggplot instead of networkD3 package and DRY the code ! :)

maurolepore reviewed (approved):
LGTM. See my comments and also ...

* Not sure what was the cause of the installation problem that Anne reported, but it seems it's not worth getting to the bottom of it. Removing and reinstalling software is a very valid way of solving an installation problem.

* I also see the annoying messages that Anne reported.

![image](https://github.com/2DegreesInvesting/tiltPlot/assets/5856545/6f2ca2bd-26d5-41bd-bb4b-656b2029443f)

Surprisingly I see them only when I run tests from the terminal but not with `devtools::test()` in the R console or from the "Test" GUI in RStudio. 

You may explore this in debugger mode and see if you can suppress it -- maybe via `as.data.frame()`. This suggests networkD3 doesn't play well with the tidyverse, which is a decent reason to consider a ggplot2 extension instead.

AnneSchoenauer reviewed (approved):
Dear Linda, 
Thanks a lot - from a analyst point of view I receive exactly the plot that I wished to have. I will make another ticket for enhancement (prettier colours, and also adding the finance_weights into the plot, achieving that the flows are not interrupted by the node but start where they ended after the node). Thanks again. 
All the best 
Anne
